// backend-aula/functions-insomnia.js
const fs = require("fs-extra");
const path = require("path");

const extrairModels = () => {
  const schemaPath = path.join(process.cwd(), "prisma/schema.prisma");
  const conteudo = fs.readFileSync(schemaPath, "utf-8");

  const modelRegex = /model\s+(\w+)\s+{([\s\S]*?)}/g;

  const models = [];
  let match;

  while ((match = modelRegex.exec(conteudo)) !== null) {
    const nomeModel = match[1];
    const bloco = match[2];

    const campos = [];

    const linhas = bloco.split("\n");

    const tiposPrimitivos = [
      "String",
      "Int",
      "Float",
      "Boolean",
      "DateTime",
      "BigInt",
      "Decimal",
    ];

    linhas.forEach((linha) => {
      linha = linha.trim();

      if (!linha || linha.startsWith("//")) return;

      const partes = linha.split(/\s+/);
      const nomeCampo = partes[0];
      const tipoCampo = partes[1].replace("?", "");

      if (!nomeCampo || !tipoCampo) return;

      // Ignorar listas (Venda[])
      if (tipoCampo.includes("[]")) return;

      // Ignorar campos de relação (Cliente, Produto, etc)
      if (!tiposPrimitivos.includes(tipoCampo)) return;

      if (nomeCampo && tipoCampo && !linha.startsWith("@")) {
        campos.push({ nome: nomeCampo, tipo: tipoCampo });
      }
    });

    models.push({ nome: nomeModel, campos });
  }

  return models;
};

function gerarBodyExemplo(campos) {
  const body = {};

  campos.forEach((campo) => {
    if (campo.nome === "id") return;

    switch (campo.tipo) {
      case "String":
        body[campo.nome] = "string";
        break;
      case "Int":
        body[campo.nome] = 0;
        break;
      case "Float":
        body[campo.nome] = 0.0;
        break;
      case "Boolean":
        body[campo.nome] = true;
        break;
      case "DateTime":
        body[campo.nome] = new Date().toISOString();
        break;
      default:
        body[campo.nome] = null;
    }
  });

  return JSON.stringify(body, null, 2);
}

function gerarInsomnia() {
  const models = extrairModels();
  const nomeProjeto = path.basename(process.cwd());

  const resources = [];

  // Workspace
  const workspace = {
    _id: "wrk_backend",
    _type: "workspace",
    name: "Backend Aula",
  };

  resources.push(workspace);

  // Pasta principal com nome do projeto
  const pastaProjetoId = `fld_${nomeProjeto}`;

  resources.push({
    _id: pastaProjetoId,
    _type: "request_group",
    parentId: "wrk_backend",
    name: nomeProjeto,
  });

  models.forEach((model) => {
    const baseName = model.nome.toLowerCase();
    const pastaModelId = `fld_${model.nome}`;

    // Subpasta do model
    resources.push({
      _id: pastaModelId,
      _type: "request_group",
      parentId: pastaProjetoId,
      name: model.nome,
    });

    const bodyExemplo = gerarBodyExemplo(model.campos);

    // GET ALL
    resources.push({
      _id: `req_${model.nome}_GET_ALL`,
      _type: "request",
      parentId: pastaModelId,
      headers: [
        {
          name: "Content-Type",
          value: "application/json",
        },
      ],
      name: `Listar`,
      method: "GET",
      url: `http://localhost:3000/${baseName}/listar`,
    });

    // GET BY ID
    resources.push({
      _id: `req_${model.nome}_GET_ID`,
      _type: "request",
      parentId: pastaModelId,
      headers: [
        {
          name: "Content-Type",
          value: "application/json",
        },
      ],
      name: `Buscar`,
      method: "GET",
      url: `http://localhost:3000/${baseName}/buscar/:id`,
    });

    // POST
    resources.push({
      _id: `req_${model.nome}_POST`,
      _type: "request",
      parentId: pastaModelId,
      name: `Cadastrar`,
      headers: [
        {
          name: "Content-Type",
          value: "application/json",
        },
      ],
      method: "POST",
      url: `http://localhost:3000/${baseName}/cadastrar`,
      body: {
        mimeType: "application/json",
        text: bodyExemplo,
      },
    });

    // PUT
    resources.push({
      _id: `req_${model.nome}_PUT`,
      _type: "request",
      parentId: pastaModelId,
      headers: [
        {
          name: "Content-Type",
          value: "application/json",
        },
      ],
      name: `Atualizar`,
      method: "PUT",
      url: `http://localhost:3000/${baseName}/atualizar/:id`,
      body: {
        mimeType: "application/json",
        text: bodyExemplo,
      },
    });

    // DELETE
    resources.push({
      _id: `req_${model.nome}_DELETE`,
      _type: "request",
      parentId: pastaModelId,
      headers: [
        {
          name: "Content-Type",
          value: "application/json",
        },
      ],
      name: `Excluir`,
      method: "DELETE",
      url: `http://localhost:3000/${baseName}/excluir/:id`,
    });
  });

  const exportJson = {
    _type: "export",
    __export_format: 4,
    __export_date: new Date().toISOString(),
    __export_source: "backend-aula",
    resources,
  };

  fs.writeFileSync(
    path.join(process.cwd(), "insomnia.json"),
    JSON.stringify(exportJson, null, 2),
  );

  console.log("Collection organizada gerada com sucesso!");
}

module.exports = {
  gerarInsomnia,
};
