// backend-aula/functions.js
const fs = require("fs-extra");
const path = require("path");

const { execSync } = require("child_process");

const criarProjeto = (projectName) => {
  //const projectName = process.argv[2];

  if (!projectName) {
    console.log("Informe o nome do projeto:");
    console.log("npx backend-aula nome-do-projeto");
    process.exit(1);
  }

  const projectPath = path.join(process.cwd(), projectName);

  fs.ensureDirSync(projectPath);
  fs.ensureDirSync(path.join(projectPath, "src"));
  fs.ensureDirSync(path.join(projectPath, "src/data"));

  fs.writeJsonSync(
    path.join(projectPath, "package.json"),
    {
      name: projectName,
      version: "1.0.0",
      main: "server.js",
      scripts: {
        dev: "node --watch server.js",
      },
    },
    { spaces: 2 },
  );

  fs.writeFileSync(
    path.join(projectPath, ".env"),
    `PORT=3000
DATABASE_URL="mysql://root@localhost:3306/mydb"
    `,
  );

  fs.writeFileSync(
    path.join(projectPath, "server.js"),
    `require('dotenv').config();
const express = require('express');
const cors = require("cors");

const app = express();
app.use(express.json());
app.use(cors());

const PORT = process.env.PORT || 3000;

app.listen(PORT, () => {
  console.log(\`Servidor rodando na porta \${PORT}\`);
});
`,
  );

  fs.writeFileSync(
    path.join(projectPath, "prisma.config.ts"),
    `// This file was generated by Prisma, and assumes you have installed the following:
// npm install --save-dev prisma dotenv
import "dotenv/config";
import { defineConfig } from "prisma/config";

export default defineConfig({
  schema: "prisma/schema.prisma",
  migrations: {
    path: "prisma/migrations",
  },
  datasource: {
    url: process.env["DATABASE_URL"],
  },
});
`,
  );

  fs.writeFileSync(
    path.join(projectPath, "src/data/prisma.js"),
    `const { PrismaClient } = require("@prisma/client");
const { PrismaMariaDb } = require("@prisma/adapter-mariadb");

const adapter = new PrismaMariaDb(process.env.DATABASE_URL);

const prisma = new PrismaClient({ adapter });

module.exports = prisma;
`,
  );

  fs.ensureDirSync(path.join(projectPath, "prisma"));

  fs.writeFileSync(
    path.join(projectPath, "prisma/schema.prisma"),
    `// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
}
`,
  );

  console.log("Instalando dependências...");

  execSync(
    "npm install express dotenv cors prisma @prisma/client @prisma/adapter-mariadb",
    { stdio: "inherit", cwd: projectPath },
  );

  console.log("Dependências instaladas!");

  console.log("Projeto criado com sucesso!");
  console.log("Agora execute:");
  console.log("cd " + projectName);
  console.log("Para inicializar o projeto execute: ");
  console.log("npm run dev");
};

const criarRecurso = (resourceName) => {
  if (!resourceName) {
    console.log("Informe o nome do recurso");
    process.exit(1);
  }

  const basePath = process.cwd();

  fs.ensureDirSync(path.join(basePath, "src/controllers"));
  fs.ensureDirSync(path.join(basePath, "src/routes"));

  const controllerPath = path.join(
    basePath,
    "src/controllers",
    `${resourceName}.controller.js`,
  );
  const routePath = path.join(
    basePath,
    "src/routes",
    `${resourceName}.routes.js`,
  );

  fs.writeFileSync(
    controllerPath,
    `const prisma = require("../data/prisma");

const cadastrar = async (req, res) => {
    const data = req.body;

    const item = await prisma.${resourceName}.create({
        data
    });

    res.json(item).status(201).end();
};

const listar = async (req, res) => {
    const lista = await prisma.${resourceName}.findMany();

    res.json(lista).status(200).end();
};

const buscar = async (req, res) => {
    const { id } = req.params;
    
    const item = await prisma.${resourceName}.findUnique({
        where: { id : Number(id) }
    });

    res.json(item).status(200).end();
};

const atualizar = async (req, res) => {
    const { id } = req.params;
    const dados = req.body;
    
    const item = await prisma.${resourceName}.update({
        where: { id : Number(id) },
        data: dados
    });

    res.json(item).status(200).end();
};

const excluir = async (req, res) => {
    const { id } = req.params;
    
    const item = await prisma.${resourceName}.delete({
        where: { id : Number(id) }
    });

    res.json(item).status(200).end();
};

module.exports = {
    cadastrar,
    listar,
    buscar,
    atualizar,
    excluir
}
`,
  );

  fs.writeFileSync(
    routePath,
    `const express = require("express");

const router = express.Router();

const { 
    cadastrar, 
    listar, 
    buscar, 
    atualizar, 
    excluir } = require("../controllers/${resourceName}.controller");

router.post("/cadastrar", cadastrar);
router.get("/listar", listar);
router.get("/buscar/:id", buscar);
router.put("/atualizar/:id", atualizar);
router.delete("/excluir/:id", excluir);

module.exports = router;
`,
  );

  atualizarServer(resourceName);

  console.log(`Recurso ${resourceName} criado com sucesso!`);
};

function atualizarServer(resourceName) {
  const serverPath = path.join(process.cwd(), "server.js");

  let conteudo = fs.readFileSync(serverPath, "utf-8");

  const importLinha = `const ${resourceName}Routes = require('./src/routes/${resourceName}.routes');\n`;

  const useLinha = `app.use('/${resourceName}', ${resourceName}Routes);\n`;

  conteudo = conteudo.replace(
    "app.use(cors());",
    `app.use(cors());\n\n${importLinha}\n${useLinha}`,
  );

  fs.writeFileSync(serverPath, conteudo);
}

const listarModels = () => {
  const schemaPath = path.join(process.cwd(), "prisma/schema.prisma");

  if (!fs.existsSync(schemaPath)) {
    console.log("Arquivo schema.prisma não encontrado.");
    return;
  }

  const conteudo = fs.readFileSync(schemaPath, "utf-8");

  const regex = /model\s+(\w+)\s+{/g;

  const models = [];
  let match;

  while ((match = regex.exec(conteudo)) !== null) {
    models.push(match[1]);
  }

  console.log("Models encontrados:");
  models.forEach((m) => console.log("- " + m));

  return models;
};

module.exports = {
  criarProjeto,
  criarRecurso,
  listarModels,
};
